### PR TITLE
fix(float): win_get_bordertext_col returning negative column number

### LINKED ISSUE #25746: Segmentation fault: floating window with centered long title
### Problem

Crash when opened floating window with centered long title.

```
build/bin/nvim --clean --headless +'luafile minimal.lua'
```

<details>
<summary>Full backtrace</summary>

```

warning: Can't open file anon_inode:[io_uring] which was expanded to anon_inode:[io_uring] during file-backed mapping note processing

warning: Can't open file anon_inode:[io_uring] which was expanded to anon_inode:[io_uring] during file-backed mapping note processing

warning: Can't open file anon_inode:[io_uring] which was expanded to anon_inode:[io_uring] during file-backed mapping note processing

warning: Can't open file anon_inode:[io_uring] which was expanded to anon_inode:[io_uring] during file-backed mapping note processing
[New LWP 10694]
[New LWP 10695]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/home/notomo/workspace/neovim/build/bin/nvim --clean --headless +luafile ./mini'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f5e061768c0 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
[Current thread is 1 (Thread 0x7f5e060d3b80 (LWP 10694))]

Thread 2 (LWP 10695):
#0  0x0000000000000000 in ?? ()
No symbol table info available.
Backtrace stopped: Cannot access memory at address 0x0

Thread 1 (Thread 0x7f5e060d3b80 (LWP 10694)):
#0  0x00007f5e061768c0 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
No symbol table info available.
#1  0x00007f5e06178c5b in ?? () from /lib/x86_64-linux-gnu/libc.so.6
No symbol table info available.
#2  0x00007f5e0617a2e2 in malloc () from /lib/x86_64-linux-gnu/libc.so.6
No symbol table info available.
#3  0x0000561d48945dc7 in try_malloc (size=7680) at /home/notomo/workspace/neovim/src/nvim/memory.c:89
        allocated_size = 7680
        ret = 0xb00000000
#4  0x0000561d48945eb5 in xmalloc (size=7680) at /home/notomo/workspace/neovim/src/nvim/memory.c:123
        ret = 0x0
#5  0x0000561d488ccd9b in grid_alloc (grid=0x561d48d3ada0 <msg_grid>, rows=24, columns=80, copy=false, valid=true) at /home/notomo/workspace/neovim/src/nvim/grid.c:789
        new_row = 999
        ngrid = {handle = 4, chars = 0x0, attrs = 0x0, vcols = 0x0, line_offset = 0x0, dirty_col = 0x0, rows = 0, cols = 0, valid = false, throttled = false, row_offset = 0, col_offset = 0, target = 0x0, blending = false, focusable = true, zindex = 0, comp_row = 0, comp_col = 0, comp_width = 0, comp_height = 0, comp_index = 0, comp_disabled = false}
        ncells = 1920
#6  0x0000561d4894c3da in msg_grid_validate () at /home/notomo/workspace/neovim/src/nvim/message.c:178
        pos = 10
        should_alloc = true
        max_rows = 23
#7  0x0000561d487f8e6a in showmode () at /home/notomo/workspace/neovim/src/nvim/drawscreen.c:933
        length = 0
        do_mode = -1
        can_show_mode = 73
        ruler_win = 0x7ffc006aac58
#8  0x0000561d487f4d6b in update_screen () at /home/notomo/workspace/neovim/src/nvim/drawscreen.c:679
        did_intro = false
        is_stl_global = false
        type = 40
        hl_changed = true
        providers = {size = 0, capacity = 4, items = 0x7ffc246aad88, init_array = {0x561d494a55b0, 0x0, 0x463e39db7d1f1200, 0x7ffc246aadb0}}
        did_one = true
#9  0x0000561d48894222 in ex_redraw (eap=0x7ffc246ab180) at /home/notomo/workspace/neovim/src/nvim/ex_docmd.c:6256
        r = 1
        p = 0
#10 0x0000561d4888ac4c in execute_cmd0 (retv=0x7ffc246aae74, eap=0x7ffc246ab180, errormsg=0x7ffc246aae78, preview=false) at /home/notomo/workspace/neovim/src/nvim/ex_docmd.c:1632
No locals.
#11 0x0000561d4888a5f6 in execute_cmd (eap=0x7ffc246ab180, cmdinfo=0x7ffc246ab2e8, preview=false) at /home/notomo/workspace/neovim/src/nvim/ex_docmd.c:1710
        errormsg = 0x0
        retv = 0
        save_cmdmod = {cmod_flags = 0, cmod_split = 0, cmod_tab = 0, cmod_filter_pat = 0x0, cmod_filter_regmatch = {regprog = 0x0, startp = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, endp = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, rm_matchcol = 0, rm_ic = false}, cmod_filter_force = false, cmod_verbose = 0, cmod_save_ei = 0x0, cmod_did_sandbox = 0, cmod_verbose_save = 0, cmod_save_msg_silent = 0, cmod_save_msg_scroll = 0, cmod_did_esilent = 0}
#12 0x0000561d487525e1 in nvim_cmd (channel_id=9223372036854775809, cmd=0x7ffc246ab430, opts=0x7ffc246ab528, err=0x7ffc246ab538) at /home/notomo/workspace/neovim/src/nvim/api/command.c:661
        save_current_sctx = {sc_sid = -3, sc_seq = 0, sc_lnum = 0}
        save_channel_id = 0
        saved_msg_list = 0x7ffc246abb38
        private_msg_list = 0x0
        ea = {arg = 0x561d494a33c6 "", args = 0x0, arglens = 0x0, argc = 0, nextcmd = 0x0, cmd = 0x561d494a33c0 "redraw", cmdlinep = 0x7ffc246ab178, cmdline_tofree = 0x0, cmdidx = CMD_redraw, argt = 17301762, skip = 0, forceit = 0, addr_count = 0, line1 = 1, line2 = 1, addr_type = ADDR_NONE, flags = 0, do_ecmd_cmd = 0x0, do_ecmd_lnum = 0, append = 0, usefilter = 0, amount = 0, regname = 0, force_bin = 0, read_edit = 0, mkdir_p = 0, force_ff = 0, force_enc = 0, bad_char = 0, useridx = 0, errmsg = 0x0, getline = 0x0, cookie = 0x0, cstack = 0x0}
        cmdinfo = {cmdmod = {cmod_flags = 0, cmod_split = 0, cmod_tab = 0, cmod_filter_pat = 0x0, cmod_filter_regmatch = {regprog = 0x0, startp = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, endp = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, rm_matchcol = 0, rm_ic = false}, cmod_filter_force = false, cmod_verbose = 0, cmod_save_ei = 0x0, cmod_did_sandbox = 0, cmod_verbose_save = 0, cmod_save_msg_silent = 0, cmod_save_msg_scroll = 0, cmod_did_esilent = 0}, magic = {file = false, bar = true}}
        cmdline = 0x561d494a33c0 "redraw"
        cmdname = 0x561d494a3f00 "redraw"
        args = {size = 0, capacity = 0, items = 0x0}
        retv = {data = 0x0, size = 0}
        p = 0x561d494a3f06 ""
        fullname = 0x561d48c7890d "redraw"
        capture_local = {ga_len = 152, ga_maxlen = 0, ga_itemsize = 0, ga_growsize = 0, ga_data = 0x0}
        save_msg_silent = 0
        save_capture_ga = 0x0
        save_msg_col = 0
#13 0x0000561d48733752 in nlua_api_nvim_cmd (lstate=0x7f5e05dc6380) at /home/notomo/workspace/neovim/build/src/nvim/auto/lua_api_c_bindings.generated.c:1607
        err = {type = kErrorTypeNone, msg = 0x0}
        err_param = 0x0
        arg2 = {output = false}
        arg1 = {is_set__cmd_ = 66, cmd = {data = 0x561d494a3ea0 "redraw", size = 6}, range = {size = 0, capacity = 0, items = 0x0}, count = 0, reg = {data = 0x0, size = 0}, bang = false, args = {size = 0, capacity = 0, items = 0x0}, magic = {size = 0, capacity = 0, items = 0x0}, mods = {size = 0, capacity = 0, items = 0x0}, nargs = {type = kObjectTypeNil, data = {boolean = false, integer = 0, floating = 0, string = {data = 0x0, size = 0}, array = {size = 0, capacity = 0, items = 0x0}, dictionary = {size = 0, capacity = 0, items = 0x0}, luaref = 0}}, addr = {type = kObjectTypeNil, data = {boolean = false, integer = 0, floating = 0, string = {data = 0x0, size = 0}, array = {size = 0, capacity = 0, items = 0x0}, dictionary = {size = 0, capacity = 0, items = 0x0}, luaref = 0}}, nextcmd = {type = kObjectTypeNil, data = {boolean = false, integer = 0, floating = 0, string = {data = 0x0, size = 0}, array = {size = 0, capacity = 0, items = 0x0}, dictionary = {size = 0, capacity = 0, items = 0x0}, luaref = 0}}}
        ret = {data = 0x7f5e05dc6380 "(\026\333\005^\177", size = 5513868316}
#14 0x0000561d48ba9096 in lj_BC_FUNCC ()
No symbol table info available.
#15 0x0000561d48b92efa in lua_pcall (L=0x561d494a52c0, nargs=<optimized out>, nresults=<optimized out>, errfunc=592) at lj_api.c:1150
        g = 0x7f5e05dc63e0
        oldh = <optimized out>
        ef = 416611828305
        status = <optimized out>
#16 0x0000561d48900c42 in nlua_pcall (lstate=0x7f5e05dc6380, nargs=0, nresults=0) at /home/notomo/workspace/neovim/src/nvim/lua/executor.c:165
        status = 0
#17 0x0000561d4890187e in nlua_exec_file (path=0x561d494a3ec8 "./minimal.lua") at /home/notomo/workspace/neovim/src/nvim/lua/executor.c:1818
        lstate = 0x7f5e05dc6380
#18 0x0000561d48903378 in ex_luafile (eap=0x7ffc246ab7e8) at /home/notomo/workspace/neovim/src/nvim/lua/executor.c:1749
No locals.
#19 0x0000561d4888ac4c in execute_cmd0 (retv=0x7ffc246ab7e4, eap=0x7ffc246ab7e8, errormsg=0x7ffc246ab8a8, preview=false) at /home/notomo/workspace/neovim/src/nvim/ex_docmd.c:1632
No locals.
#20 0x0000561d48886bd1 in do_one_cmd (cmdlinep=0x7ffc246abb98, flags=11, cstack=0x7ffc246abba0, fgetline=0x0, cookie=0x0) at /home/notomo/workspace/neovim/src/nvim/ex_docmd.c:2291
        errormsg = 0x0
        save_reg_executing = 0
        save_pending_end_reg_executing = false
        ea = {arg = 0x561d494a3ec8 "./minimal.lua", args = 0x0, arglens = 0x0, argc = 0, nextcmd = 0x0, cmd = 0x561d494a3ec0 "luafile ./minimal.lua", cmdlinep = 0x7ffc246abb98, cmdline_tofree = 0x0, cmdidx = CMD_luafile, argt = 17301661, skip = 0, forceit = 0, addr_count = 0, line1 = 1, line2 = 1, addr_type = ADDR_LINES, flags = 0, do_ecmd_cmd = 0x0, do_ecmd_lnum = 0, append = 0, usefilter = 0, amount = 0, regname = 0, force_bin = 0, read_edit = 0, mkdir_p = 0, force_ff = 0, force_enc = 0, bad_char = 0, useridx = 0, errmsg = 0x0, getline = 0x0, cookie = 0x0, cstack = 0x7ffc246abba0}
        save_cmdmod = {cmod_flags = 0, cmod_split = 0, cmod_tab = 0, cmod_filter_pat = 0x0, cmod_filter_regmatch = {regprog = 0x0, startp = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, endp = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, rm_matchcol = 0, rm_ic = false}, cmod_filter_force = false, cmod_verbose = 0, cmod_save_ei = 0x0, cmod_did_sandbox = 0, cmod_verbose_save = 0, cmod_save_msg_silent = 0, cmod_save_msg_scroll = 0, cmod_did_esilent = 0}
        after_modifier = 0x561d494a3ec0 "luafile ./minimal.lua"
        cmd = 0x561d494a3ec0 "luafile ./minimal.lua"
        p = 0x561d494a3ec7 " ./minimal.lua"
        ni = 0
        retv = 0
#21 0x0000561d48884682 in do_cmdline (cmdline=0x7ffc246acc49 "luafile ./minimal.lua", fgetline=0x0, cookie=0x0, flags=11) at /home/notomo/workspace/neovim/src/nvim/ex_docmd.c:590
        current_line_before = 0
        recursive = 1
        call_depth = 1
        next_cmdline = 0x561d494a3ec0 "luafile ./minimal.lua"
        cmdline_copy = 0x561d494a3ec0 "luafile ./minimal.lua"
        used_getline = false
        msg_didout_before_start = false
        count = 1
        did_inc = false
        retval = 1
        cstack = {cs_flags = {0 <repeats 50 times>}, cs_pending = '\000' <repeats 49 times>, cs_pend = {csp_rv = {0x0 <repeats 50 times>}, csp_ex = {0x0 <repeats 50 times>}}, cs_forinfo = {0x0 <repeats 50 times>}, cs_line = {0 <repeats 50 times>}, cs_idx = -1, cs_looplevel = 0, cs_trylevel = 0, cs_emsg_silent_list = 0x0, cs_lflags = 0}
        lines_ga = {ga_len = 0, ga_maxlen = 0, ga_itemsize = 16, ga_growsize = 10, ga_data = 0x0}
        current_line = 0
        fname = 0x0
        breakpoint = 0x0
        dbg_tick = 0x0
        debug_saved = {trylevel = 0, force_abort = 0, caught_stack = 0x0, vv_exception = 0x0, vv_throwpoint = 0x0, did_emsg = 0, got_int = 0, did_throw = false, need_rethrow = 0, check_cstack = 0, current_exception = 0x0}
        initial_trylevel = 0
        saved_msg_list = 0x0
        private_msg_list = 0x0
        cmd_getline = 0x0
        cmd_cookie = 0x0
        cmd_loop_cookie = {lines_gap = 0x0, current_line = 0, repeating = 0, getline = 0x0, cookie = 0x0}
        real_cookie = 0x0
        getline_is_func = 0
#22 0x0000561d48885411 in do_cmdline_cmd (cmd=0x7ffc246acc49 "luafile ./minimal.lua") at /home/notomo/workspace/neovim/src/nvim/ex_docmd.c:289
No locals.
#23 0x0000561d4872b307 in exe_commands (parmp=0x7ffc246ac240) at /home/notomo/workspace/neovim/src/nvim/main.c:1914
        i = 0
#24 0x0000561d487277f5 in main (argc=5, argv=0x7ffc246ac498) at /home/notomo/workspace/neovim/src/nvim/main.c:575
        fname = 0x0
        params = {argc = 5, argv = 0x7ffc246ac498, use_vimrc = 0x561d48c591aa "NONE", clean = true, n_commands = 2, commands = {0x7ffc246acc49 "luafile ./minimal.lua", 0x7ffc246acc60 "quitall!", 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, cmds_tofree = "\000\000\000\000\000\000\000\000\000", n_pre_commands = 0, pre_commands = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, luaf = 0x0, lua_arg0 = -1, edit_type = 0, tagname = 0x0, use_ef = 0x0, input_istext = false, no_swap_file = 0, use_debug_break_level = -1, window_count = 1, window_layout = 0, diff_mode = 0, listen_addr = 0x0, remote = 0, server_addr = 0x0, scriptin = 0x0, scriptout = 0x0, scriptout_append = false, had_stdin_file = false}
        cwd = 0x0
        has_term = true
        use_builtin_ui = false
        remote_ui = false
        use_remote_ui = false
        err = {type = kErrorTypeNone, msg = 0x0}
        o = {type = kObjectTypeNil, data = {boolean = false, integer = 0, floating = 0, string = {data = 0x0, size = 0}, array = {size = 0, capacity = 0, items = 0x0}, dictionary = {size = 0, capacity = 0, items = 0x0}, luaref = 0}}
        vimrc_none = true
```

</details>

### Steps to reproduce

minimal.lua
```lua
vim.api.nvim_open_win(0, true, {
  width = 10,
  height = 10,
  relative = "editor",
  row = 10,
  col = 10,
  focusable = true,
  external = false,
  style = "minimal",
  title = ("a"):rep(14),
  title_pos = "center",
  border = "single",
})
vim.cmd.redraw()
vim.cmd.close()
```

```
build/bin/nvim --clean --headless +'luafile minimal.lua'
zsh: segmentation fault (core dumped)
```

### Expected behavior

No crash.

### Neovim version (nvim -v)

0.10.0-dev-1399+g3a3e02512

### Vim (not Nvim) behaves the same?

N/A (nvim_open_win)

### Operating system/version

Ubuntu 22.04

### Terminal name/version

wezterm 20230726-161042-3aa7da61

### $TERM environment variable

xterm-256color

### Installation

build from repo

### PR DESCRIPTION
**Pull request by ghost**
Problem:
  `win_get_bordertext_col` returns column < 1 for right or center
  aligned text, if its length is more than window width.

Solution:
  Return max(resulting_column, 1)

fixes #25746

### CODE REVIEW

File: test/functional/ui/float_spec.lua [hunk 1]
```diff
@@ -2034,6 +2034,61 @@ describe('float window', function()
...
+          ^                                        |
+          {0:~                                       }|
+          {0:~                                       }|
+          {0:~                                       }|
+          {0:~                                       }|
+          {0:~                                       }|
+        ## grid 3
+                                                  |
+        ## grid 4
+          {5:╔}{11:abcdefghi}{5:╗}|
```
Comment (glepnir) [inline test/functional/ui/float_spec.lua:2066]: This preserves the first half. truncate the title before render it also fixes it. like the statusline `<text` keep the second half.  not sure which is good way :thinking:

Comment (nwounkn) [reply test/functional/ui/float_spec.lua:2066]: another way is to skip text while column < 1 - truncates center aligned text from both sides (adapted from `grid_line_puts`):
```diff
diff --git a/src/nvim/drawscreen.c b/src/nvim/drawscreen.c
index adbe0f473..896d0c4a6 100644
--- a/src/nvim/drawscreen.c
+++ b/src/nvim/drawscreen.c
@@ -725,7 +725,17 @@ static void win_redr_bordertext(win_T *wp, VirtText vt, int col)
     if (text == NULL) {
       break;
     }
-    col += grid_line_puts(col, text, -1, attr);
+    while (col < 1 && *text != NUL) {
+      int u8cc[MAX_MCO];
+      int mbyte_blen = utfc_ptr2len(text);
+      int u8c = utfc_ptr2char(text, u8cc);
+      int mbyte_cells = utf_char2cells(u8c);
+      col += MIN(mbyte_cells, 2);
+      text += mbyte_blen;
+    }
+    if (col >= 1) {
+      col += grid_line_puts(col, text, -1, attr);
+    }
   }
 }
```

Comment (glepnir) [reply test/functional/ui/float_spec.lua:2066]: Yes. I prefer to truncate the title. But I also confused. Can't the titile length and window width be calculated locally before setting them.  But in any case should avoid crash in core.

Comment (nwounkn) [reply test/functional/ui/float_spec.lua:2066]: The title length is already calculated, its just the alignment makes the text start outside the window. The truncated title needs to be calculated every time title, title_pos, and width change.

**Pull request merged**